ahcer/src/app/episode-report/episode-report.component.ts

```diff
@@ -0,0 +1,283 @@
+        }
+        else {
+          this.loadingPatient = false;
+        }
+      })
+  }
+
+  @ViewChild(MatSort) sort: MatSort;
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+
+  sortingDataAccessor = (item: Episode, property: string) => {
+    switch (property) {
+      case 'startTime':
+        return item.startTime.seconds;
+      case 'endTime':
+        return item.endTime?.seconds;
+      case 'duration':
+        return (item.endTime)? item.endTime.seconds - item.startTime.seconds: null;
+      case 'symptoms':
+        return this.displaySymptomsString(item.symptoms);
+      default:
+        return item[property];
+    }
+  }
+
+  reloadDataSource() {
+    this.dataSource.data = this.episodes;
+    this.dataSource.sort = this.sort;
+    this.dataSource.sortingDataAccessor = this.sortingDataAccessor;
+    this.dataSource.paginator = this.paginator;
+  }
+
+  @HostListener('window:beforeprint',['$event'])
+  onBeforePrint(event){
+    this.dataSource.data = this.episodes;
+    this.dataSource.sort = this.sort;
+    this.dataSource.sortingDataAccessor = this.sortingDataAccessor;
+    this.dataSource.paginator = null;
+  }
+
+  @HostListener('window:afterprint', ['$event'])
+  onAfterPrint(event) {
+    this.reloadDataSource();
+  }
+
+  ngAfterViewInit() {
+    this.reloadDataSource();
+  }
+
+  loadEpisodes(patientId: string) {
+    this.loadingEpisodes = true;
+    this.episodeService.getAllEpisodesByPatient(patientId, 'desc')
+      .pipe(
+        finalize(() => {
+          this.loadingEpisodes = false;
+          this.episodes_count = this.episodes.length;
+          this.reloadDataSource();
+        })
+      )
+      .subscribe(
+        episodes => {this.episodes = episodes}
+      );
+  }
+
+  changePatient(patientId: string) {
+    this.usersService.changeLastViewedPatient(patientId)
+      .subscribe(()=> {
+        this.loadPatient(patientId)
+      })
+  }
+
+  loadPatient(patientId: string) {
+    if(patientId == this.currentPatient?.id) {
+      return
+    }
+    this.patientsService.getPatientById(patientId)
+      .pipe(
+        finalize(()=> this.loadingPatient = false)
+      )
+      .subscribe(patient => {
+        this.currentPatient = patient;
+        this.loadEpisodes(patientId);
+        this.loadRescueMeds(patientId);
+      })
+  }
+
+  loadRescueMeds(patientId: string) {
+    this.loadingRescueMeds = true;
+    this.medicationService.getMedicationsByType(patientId, true,  false, false)
+      .pipe(
+        finalize(()=> this.loadingRescueMeds=false)
+      ).subscribe(medications => {
+        for(let med of medications) {
+          this.rescueMedNames[med.id] = med.name;
+        }
+      })
+  }
+
+  calculateDuration(startTime: Timestamp, endTime: Timestamp) {
+    return this.episodeService.calculateDuration(startTime, endTime);
+  }
+
+  //needs rework when symptom management is implemented
+  displaySymptomsString(symptoms: Episode["symptoms"], showFullList: boolean = false): string {
```

Comment:
I think this is good enough for now...but I'm sure we could find a way to simplify this.